### PR TITLE
10 service category

### LINKED ISSUE #10: Service: Category
Create a service to read/write/list all categories. Is a wrapper for calls to DataStore API.

### PR DESCRIPTION
**Pull request by sschoetz**
Implement category service. AWS Back-end requires login functionality to be merged first.
Closes #10.

### CODE REVIEW

File: ctt-frontend/src/app/services/aws-category.service.spec.ts [hunk 1]
```diff
@@ -1,16 +1,217 @@
...
-//   let service: AwsCategoryService;
+describe('AwsCategoryService', () => {
+  let service: AwsCategoryService;
 
-//   beforeEach(() => {
-//     TestBed.configureTestingModule({});
-//     service = TestBed.inject(AwsCategoryService);
-//   });
+  beforeEach(() => {
+    TestBed.configureTestingModule({});
```
Comment (walli545) [inline ctt-frontend/src/app/services/aws-category.service.spec.ts]: You can remove the `TestBed` setup from `beforeEach` and the `it should be created` test. Just create the `apiMock` and `sut` objects for each test in the `beforeEach` method manually and store them as global variables (like `service`).
Then you only have the `apiMock.configure(...)`  call in each test and you lose some lines in the test code.

File: ctt-frontend/src/app/services/aws-category.service.spec.ts [hunk 2]
```diff
@@ -1,16 +1,217 @@
...
+        .setup((x) =>
+          x.CreateCategory(TypeMoq.It.is((y) => (y as APICreateInput) !== null))
+        )
+        .returns((y) => Promise.resolve(DummyData.createMutation));
+      const sut = new AwsCategoryService(apiMock.object);
+
+      await expectAsync(sut.create(DummyData.categoryInput)).toBeResolvedTo(
+        DummyData.categoryValue
+      );
+      expect(
```
Comment (walli545) [inline ctt-frontend/src/app/services/aws-category.service.spec.ts]: Also the `expect(...)` wrapping around the `apiMock.verify(...)` call is redundant and can be removed. `verify` throws when no call is found and so the test fails.

File: ctt-frontend/src/app/services/aws-category.service.ts [hunk 3]
```diff
@@ -1,56 +1,113 @@
...
-      name: 'sleep',
-      id: 'a',
-      reminderInterval: 10,
-      excludeFromStatistics: [StatisticType.absoluteTime],
-    });
+    const i: APICreateInput = {
+      name: input.name,
+      color: input.color,
+      reminderInterval: input.reminderInterval,
+      excludeFromStatistics: input.excludeFromStatistics.map((v) =>
```
Comment (walli545) [inline ctt-frontend/src/app/services/aws-category.service.ts:22]: Maybe add private methods `convertStatisticTypeToApiStatisticType()` and  `convertApiStatisticTypeToStatisticType()`

File: ctt-frontend/src/app/services/aws-category.service.ts [hunk 4]
```diff
@@ -1,56 +1,113 @@
...
+          ? APIStatisticType.AbsoluteTime
+          : APIStatisticType.RelativeTime
+      ),
+    } as APICreateInput;
+    return this.api.CreateCategory(i).then(
+      (result) =>
+        Promise.resolve({
+          color: result.color,
+          id: result.id,
+          name: result.name,
```
Comment (walli545) [inline ctt-frontend/src/app/services/aws-category.service.ts:31]: I think also private methods for converting an `Category` to an `ApiCategory` and the other way make sense.

**Review by walli545 — CHANGES_REQUESTED** [review]
Looks good! Nice that you also included tests! Please look at my improvement suggestions. Also, I would wait to merge this until we can test the functionality when signed in.

**Review by AndHager — APPROVED** [review]

**Review by walli545 — APPROVED** [review]

**Pull request merged**